### PR TITLE
[fix] #4140: Fix registration of new peer

### LINKED ISSUE #4140: [BUG] Fix new peer registration

### PR DESCRIPTION
**Pull request by mversic**
## Description

* Fixed new peer registration
* Fixed peer restart

### Linked issue

Closes #4140

### Checklist

- [ ] I've read `CONTRIBUTING.md`
- [ ] I've used the standard signed-off commit format (or will squash just before merging)
- [ ] All applicable CI checks pass (or I promised to make them pass later)
- [ ] (optional) I've written unit tests for the code changes
- [ ] I replied to all comments after code review, marking all implemented changes with thumbs up

### CODE REVIEW

File: client/tests/integration/connected_peers.rs [hunk 1]
```diff
@@ -1,15 +1,15 @@
...
-    data_model::{
-        parameter::{default::MAX_TRANSACTIONS_IN_BLOCK, ParametersBuilder},
-        peer::Peer as DataModelPeer,
-        prelude::*,
-    },
+use iroha_client::{client::Client, data_model::peer::Peer as DataModelPeer};
+use iroha_data_model::{
+    isi::{RegisterExpr, UnregisterExpr},
+    IdBox,
 };
```
Comment (Erigara) [inline client/tests/integration/connected_peers.rs:8]: Why have you changed this part?
Afaik you previously made this change, to have control over `iroha_data_model` export.

File: config/src/sumeragi.rs [hunk 2]
```diff
@@ -94,11 +94,16 @@ impl ConfigurationProxy {
...
         let peer_id = self
             .peer_id
-            .clone()
+            .as_ref()
             .expect("Insertion of `self` as `trusted_peers` implies that `peer_id` field should be initialized");
-        self.trusted_peers = Some(TrustedPeers {
-            peers: unique_vec![peer_id],
-        });
+        self.trusted_peers = if let Some(mut trusted_peers) = self.trusted_peers.take() {
+            trusted_peers.peers.push(peer_id.clone());
```
Comment (Erigara) [inline config/src/sumeragi.rs:100]: What if self already present in `trusted_peers`?

File: docker-compose.single.yml
Comment (Erigara) [inline docker-compose.single.yml:1]: Afaik this compose files are auto generated, aren't they? 
Should you update script as well?

**Comment by Erigara** [comment]
I wounder how we should deal with more general case when whole topology changed.

Step by step: 
1. Network is started with peer set A
2. After some height peer set A is replaced with peer set B (A intersection B = empty set)
3. New peer is added with trusted peers peers from set B 
4. Peer get block and update it's topology to set A (because at this height only peers from topology A were present)

File: client/tests/integration/connected_peers.rs (hunk 1 above)
Comment (mversic) [reply client/tests/integration/connected_peers.rs:8]: accidentally, will change back

File: config/src/sumeragi.rs (hunk 2 above)
Comment (mversic) [reply config/src/sumeragi.rs:100]: no problem, it's backed by `unique_vec` so it will not add a duplicate

File: docker-compose.single.yml
Comment (mversic) [reply docker-compose.single.yml:1]: working on it now

File: config/src/sumeragi.rs (hunk 2 above)
Comment (Erigara) [reply config/src/sumeragi.rs:100]: oh indeed

**Comment by mversic** [comment]
> I wounder how we should deal with more general case when whole topology changed.

good point. We should discuss the best solution in a separate ticket

File: docker-compose.single.yml
Comment (mversic) [reply docker-compose.single.yml:1]: no, wait. Only `dev` compose files are auto generated. This one is not

**Review by Erigara — APPROVED** [review]

File: core/src/sumeragi/mod.rs [hunk 3]
```diff
@@ -264,40 +286,21 @@ impl SumeragiHandle {
...
         for block in block_iter_except_last {
-            let block = ValidBlock::validate(Clone::clone(&block), &current_topology, &mut wsv)
-                .expect("Kura blocks should be valid")
-                .commit(&current_topology)
-                .expect("Kura blocks should be valid");
-            wsv.apply_without_execution(&block).expect(
-                "Block application in init should not fail. \
-                 Blocks loaded from kura assumed to be valid",
-            );
+            current_topology = Self::replay_block(&block, &mut wsv, &current_topology);
```
Comment (mversic) [inline core/src/sumeragi/mod.rs:296]: I don't understand how this worked previously. Did it?

Comment (mversic) [reply core/src/sumeragi/mod.rs:296]: lol, this was broken. I fixed and improved the test

File: client/tests/integration/offline_peers.rs [hunk 4]
```diff
@@ -43,3 +40,41 @@ fn genesis_block_is_committed_with_some_offline_peers() -> Result<()> {
...
+    let public_key = key_pair.public_key().clone();
+    let peer_id = PeerId::new(&address, &public_key);
+    let register_peer = Register::peer(DataModelPeer::new(peer_id));
+
+    // Wait for some time to allow peers to connect
+    client.submit_blocking(register_peer)?;
+    std::thread::sleep(pipeline_time * 2);
+
+    // Make sure status hasn't change
+    check_status(&peer_clients, 2);
```
Comment (Erigara) [inline client/tests/integration/offline_peers.rs:66]: But status has changed from `1` to `2`, no?

Comment (mversic) [reply client/tests/integration/offline_peers.rs:66]: that's the block height. One block was produced

Comment (mversic) [reply client/tests/integration/offline_peers.rs:66]: the comment is misleading

File: config/src/iroha.rs [hunk 5]
```diff
@@ -144,11 +144,8 @@ impl ConfigurationProxy {
                     message: "Torii config should have at least `p2p_addr` provided for sumeragi finalisation",
                 });
             }
-            // Finally, if trusted peers were not supplied, we can fall back to inserting itself as
-            // the only trusted one
-            if sumeragi_proxy.trusted_peers.is_none() {
```
Comment (VAmuzing) [inline config/src/iroha.rs:149]: Is the check not necessary and `trusted_peers` is always `None`?

Comment (mversic) [reply config/src/iroha.rs:149]: it's not always `None` check the `insert_self_as_trusted_peers`

Comment (VAmuzing) [reply config/src/iroha.rs:149]: oh, I missed the changes in the method.
Looked at my local version

**Review by VAmuzing — APPROVED** [review]

**Pull request merged**